server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.realtime.firehose;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.dataformat.smile.SmileFactory;
+import com.fasterxml.jackson.dataformat.smile.SmileGenerator;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Ordering;
+import com.google.common.collect.Sets;
+import com.google.common.io.CountingOutputStream;
+import com.google.common.io.Files;
+import com.metamx.emitter.EmittingLogger;
+import io.druid.data.input.Firehose;
+import io.druid.data.input.FirehoseFactory;
+import io.druid.data.input.InputRow;
+import io.druid.data.input.Row;
+import io.druid.data.input.Rows;
+import io.druid.data.input.impl.InputRowParser;
+import io.druid.jackson.DefaultObjectMapper;
+import io.druid.java.util.common.guava.CloseQuietly;
+import io.druid.java.util.common.parsers.ParseException;
+import io.druid.utils.Runnables;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Creates a wrapper firehose that writes from another firehose to disk and then serves nextRow() from disk. Useful for
+ * tasks that require multiple passes through the data to prevent multiple remote fetches. Also has support for
+ * retrying fetches if the underlying firehose throws an exception while the local cache is being generated.
+ */
+public class ReplayableFirehoseFactory implements FirehoseFactory<InputRowParser>, Closeable
+{
+  private static final EmittingLogger log = new EmittingLogger(ReplayableFirehoseFactory.class);
+  private static final boolean DELETE_TEMP_FILES = true;
+  private static final boolean DEFAULT_REPORT_PARSE_EXCEPTIONS = false;
+  private static final int DEFAULT_MAX_TEMP_FILE_SIZE = 250000000;
+  private static final int DEFAULT_READ_FIREHOSE_RETRIES = 3;
+
+  private final FirehoseFactory delegateFactory;
+  private final boolean reportParseExceptions;
+  private final int maxTempFileSize;
+  private final int readFirehoseRetries;
+
+  private ReplayableFirehose firehose;
+
+  @JsonCreator
+  public ReplayableFirehoseFactory(
```

Comment:
@gianm cannot find such statements in https://github.com/druid-io/druid/blob/d981a2aa02aedfdf18e0bed5848e704bd9597563/docs/content/development/versioning.md.

It seems strange to me, because it's inhibiting a good practice of documenting code.